x-pack/functionbeat/manager/gcp/template_builder.go

```diff
@@ -0,0 +1,195 @@
+
+	return &functionData{
+		raw:      raw,
+		function: d.cloudFunction(name, fn.Config()),
+	}, nil
+}
+
+func findFunction(p provider.Provider, name string) (installer, error) {
+	fn, err := p.FindFunctionByName(name)
+	if err != nil {
+		return nil, err
+	}
+
+	function, ok := fn.(installer)
+	if !ok {
+		return nil, errors.New("incompatible type received, expecting: 'functionManager'")
+	}
+
+	return function, nil
+}
+
+func (d *defaultTemplateBuilder) cloudFunction(name string, config *fngcp.FunctionConfig) *cloudfunctions.CloudFunction {
+	fnName := fmt.Sprintf(functionName, d.gcpConfig.ProjectID, d.gcpConfig.Location, name)
+	sourceArchiveURL := fmt.Sprintf(archiveURL, d.gcpConfig.FunctionStorage, name)
+
+	return &cloudfunctions.CloudFunction{
+		Name:        fnName,
+		Description: config.Description,
+		EntryPoint:  config.EntryPoint(),
+		EnvironmentVariables: map[string]string{
+			"ENABLED_FUNCTIONS": name,
+			"BEAT_STRICT_PERMS": "false",
+		},
+		EventTrigger: &cloudfunctions.EventTrigger{
+			EventType: config.Trigger.EventType,
+			Resource:  config.Trigger.Resource,
+			Service:   config.Trigger.Service,
+		},
+		Labels:              config.Labels,
+		MaxInstances:        int64(config.MaxInstances),
+		Runtime:             runtime,
+		ServiceAccountEmail: config.ServiceAccountEmail,
+		SourceArchiveUrl:    sourceArchiveURL,
+		Timeout:             config.Timeout.String(),
+		VpcConnector:        config.VPCConnector,
+	}
+}
+
+// RawTemplate returns the JSON to POST to the endpoint.
+func (d *defaultTemplateBuilder) RawTemplate(name string) (string, error) {
+	fn, err := findFunction(d.provider, name)
+	if err != nil {
+		return "", err
+	}
+	config := fn.Config()
+
+	properties := common.MapStr{
+		"codeLocation":     "pkg/" + fn.Name(),
+		"codeBucket":       d.gcpConfig.FunctionStorage,
+		"codeBucketObject": "functionbeat.zip",
+		"location":         d.gcpConfig.Location,
+		"runtime":          runtime,
+		"entryPoint":       config.EntryPoint(),
+		"eventTrigger":     config.Trigger,
+		"environmentVariables": common.MapStr{
+			"ENABLED_FUNCTIONS": name,
+			"BEAT_STRICT_PERMS": false,
+		},
+	}
+
+	if config.Timeout > 0*time.Second {
+		properties["timeout"] = config.Timeout.String()
+	}
+	if config.MemorySize != "" {
+		properties["availableMemoryMb"] = config.MemorySize
+	}
+	if len(config.ServiceAccountEmail) > 0 {
+		properties["serviceAccountEmail"] = config.ServiceAccountEmail
+	}
+	if len(config.Labels) > 0 {
+		properties["labels"] = config.Labels
+	}
+	if config.MaxInstances > 0 {
+		properties["maxInstances"] = config.MaxInstances
+	}
+	if len(config.VPCConnector) > 0 {
+		properties["vpcConnector"] = config.VPCConnector
+	}
+
+	output := common.MapStr{
+		"resources": []common.MapStr{
+			common.MapStr{
+				"name":       fmt.Sprintf(functionName, d.gcpConfig.ProjectID, d.gcpConfig.Location, name),
+				"type":       "cloudfunctions.v1beta2.function", // TODO
```

Comment:
Please add a comment to TODO so it is clear for everyone what the comment is about. e.g. `TODO: update to v1 (or intermittent versions) once new API version becomes available`.

Reply:
New users should use `v1` instead of `v1beta`, so i am changing it.